File: api/app/measure/framing_bench.py
"""Banc de mesure du cadrage : que coûte une photo prise à main levée ?

**Pourquoi ce banc existe.** Le pipeline découpe l'illustration à une position
fixe dans le plus grand rectangle aux proportions d'une carte, centré dans la
photo. Il suppose donc que la carte remplit ce rectangle exactement. Le premier
test terrain a mesuré la tolérance de cette hypothèse — 2 à 3 % — sans que
personne ne sache ce qu'une vraie photo lui inflige. Ce banc le chiffre, et
servira de point de comparaison à toute détection de bords : sans lui, on
saurait qu'on a écrit du code, pas qu'on a amélioré quoi que ce soit.

**Les photos sont synthétiques, et c'est un choix.** Une photo réelle porte sa
vérité terrain dans la tête de celui qui l'a prise ; une photo composée la porte
dans ses paramètres. On sait exactement de combien la carte a été décalée,
tournée, inclinée — donc à quel écart correspond quel échec. Le protocole
reprend celui déjà employé pour l'étalement (`docs/spread-detection.md`).

**Ce que le banc ne mesure pas** : le flou de bougé, la mise au point ratée, les
reflets sur un protège-carte. Ils dégradent l'empreinte elle-même, pas le
cadrage, et le pipeline leur a déjà été confronté — 120 cartes dégradées,
reconnues à 98 % dans le pire régime.

Usage :
    python -m app.measure.framing_bench            # 40 cartes, tous les régimes
    python -m app.measure.framing_bench --cards 12 # échantillon plus petit
"""

from __future__ import annotations

import argparse
import io
import math
import random
from dataclasses import dataclass

import httpx
import numpy as np
import psycopg
from PIL import Image

from app.config import SupabaseConfig
from app.vision.card_bounds import find_card, sample_art
from app.vision.dhash import dhash, from_signed_64, hamming_distance

USER_AGENT = "DeckHand/0.1 (https://github.com/Lelio88/DeckHand)"

#: Seuil de confiance du scan, en bits. Au-delà, la carte est « perdue » : ce
#: n'est pas une erreur silencieuse, l'application dit son doute — mais elle ne
#: reconnaît rien.
CONFIDENCE = 12

#: Proportions d'une carte Magic, 63 × 88 mm. Jumeau de `card_framing.dart`.
CARD_ASPECT = 63 / 88

#: Gabarit du cadre moderne, jumeau de `art_box.dart`. Le cadre ancien existe
#: aussi ; le banc s'en tient au moderne, qui couvre les cartes tirées ici.
ART_BOX = (0.080, 0.120, 0.920, 0.550)


@dataclass(frozen=True)
class Shot:
    """Un régime de prise de vue, décrit par ce que la main fait de travers."""

    name: str
    #: Marge de table autour de la carte, en fraction de sa hauteur.
    margin: float
    #: Décalage du centre, en fraction de la largeur de la carte.
    offset: float
    #: Rotation, en degrés.
    rotation: float

    def label(self) -> str:
        return (
            f"{self.name:22} marge {self.margin:4.0%}  "
            f"décalage {self.offset:4.0%}  rotation {self.rotation:4.1f}°"
        )


#: Du cadrage parfait — que personne n'atteint — au cadrage négligent.
#: Les valeurs intermédiaires encadrent ce qu'une main produit réellement :
#: quelques pour cent de marge, un ou deux degrés de travers.
REGIMES: tuple[Shot, ...] = (
    Shot("parfait", 0.00, 0.00, 0.0),
    Shot("soigné", 0.03, 0.01, 0.5),
    Shot("ordinaire", 0.08, 0.03, 2.0),
    Shot("à la volée", 0.15, 0.06, 5.0),
    Shot("négligent", 0.25, 0.10, 9.0),
)


def sample_prints(limit: int, salt: str = "cadrage") -> list[tuple[str, str, int]]:
    """Impressions tirées au hasard, avec leur empreinte de référence.

    Seules celles qui portent une empreinte comptent : c'est elle qui fait
    office de vérité terrain, puisque c'est elle que l'application cherchera.

    **Le tirage est reproductible, et c'est vital.** Un `ORDER BY random()`
    donnerait un échantillon différent à chaque exécution : deux mesures ne
    seraient plus comparables, et l'écart entre deux versions du code se
    confondrait avec l'écart entre deux paquets de cartes. Le hachage de
    l'identifiant fixe l'ordre une fois pour toutes.
    """
    config = SupabaseConfig.load()
    query = """
        SELECT p.scryfall_id::text, p.art_crop_url, a.dhash
        FROM public.card_prints p
        JOIN public.art_hashes a ON a.scryfall_id = p.scryfall_id
        JOIN public.cards c ON c.oracle_id = p.oracle_id
        WHERE c.game = 'magic'
          AND c.layout = 'normal'
          AND p.lang = 'en'
          AND p.released_at >= '2004-01-01'
        ORDER BY md5(p.scryfall_id::text || %s)
        LIMIT %s
    """
    with psycopg.connect(config.db_url, connect_timeout=30) as conn:
        with conn.cursor() as cur:
            cur.execute(query, (salt, limit))
            return [(sid, url, from_signed_64(h)) for sid, url, h in cur.fetchall()]


def fetch_card_image(art_crop_url: str, client: httpx.Client) -> Image.Image | None:
    """Carte entière, obtenue en substituant la taille dans l'URL.

    Scryfall sert la même image sous plusieurs tailles au même chemin ; passer
    d'`art_crop` à `normal` évite un appel à l'API pour retrouver l'URL.
    """
    url = art_crop_url.replace("/art_crop/", "/normal/")
    try:
        response = client.get(url, timeout=30)
        response.raise_for_status()
        return Image.open(io.BytesIO(response.content)).convert("RGB")
    except (httpx.HTTPError, OSError):
        return None


def table_background(width: int, height: int, rng: random.Random) -> Image.Image:
    """Fond de table texturé.

    Un aplat uni rendrait la détection triviale et le banc menteur : c'est la
    texture qui met les approches en difficulté, comme l'ont montré les mesures
    d'étalement.
    """
    base = np.full((height, width, 3), (168, 150, 124), dtype=np.int16)
    grain = np.array(
        rng.choices(range(-14, 15), k=height * width), dtype=np.int16
    ).reshape(height, width, 1)
    # Un dégradé diagonal imite un éclairage inégal, l'autre écueil connu.
    gradient = np.linspace(-18, 18, width, dtype=np.int16).reshape(1, width, 1)
    canvas = np.clip(base + grain + gradient, 0, 255).astype(np.uint8)
    return Image.fromarray(canvas, mode="RGB")


def compose(card: Image.Image, shot: Shot, rng: random.Random) -> Image.Image:
    """Photo synthétique : la carte posée sur une table, vue de travers."""
    card_height = 900
    card_width = round(card_height * CARD_ASPECT)
    card = card.resize((card_width, card_height), Image.LANCZOS)

    margin = round(card_height * shot.margin)
    photo_width = card_width + 2 * margin
    photo_height = card_height + 2 * margin
    photo = table_background(photo_width, photo_height, rng)

    if shot.rotation:
        # **En RGBA, sans quoi le banc mesure un artefact.** `rotate` remplit
        # les coins libérés par la rotation avec la couleur de fond — noire sur
        # une image RGB. Ce losange noir ceignant la carte est exactement ce
        # qu'un masque de carte cherche, et la détection trouvait alors les
        # coins du losange au lieu de ceux de la carte.
        card = card.convert("RGBA").rotate(
            shot.rotation, resample=Image.BICUBIC, expand=True
        )

    dx = round(card_width * shot.offset)
    dy = round(card_height * shot.offset * CARD_ASPECT)
    x = (photo_width - card.width) // 2 + dx
    y = (photo_height - card.height) // 2 + dy
    photo.paste(card, (x, y), card if card.mode == "RGBA" else None)

    # La compression est celle d'un téléphone, pas celle d'un scanner.
    buffer = io.BytesIO()
    photo.save(buffer, format="JPEG", quality=78)
    return Image.open(buffer).convert("RGB")


def crop_to_card_frame(photo: Image.Image) -> Image.Image:
    """Jumeau de `cropToCardFrame` : plus grand rectangle 63:88, centré."""
    width, height = photo.size
    crop_width = width
    crop_height = round(width / CARD_ASPECT)
    if crop_height > height:
        crop_height = height
        crop_width = round(height * CARD_ASPECT)
    left = (width - crop_width) // 2
    top = (height - crop_height) // 2
    return photo.crop((left, top, left + crop_width, top + crop_height))


def crop_art(card: Image.Image) -> Image.Image:
    """Zone d'illustration, aux proportions du gabarit moderne."""
    width, height = card.size
    left, top, right, bottom = ART_BOX
    return card.crop(
        (
            round(left * width),
            round(top * height),
            round(right * width),
            round(bottom * height),
        )
    )


def measure(
    cards: int, seed: int = 20260810, detect: bool = False
) -> tuple[dict[str, list[int]], int]:
    """Distance de l'empreinte à sa référence, par régime de prise de vue.

    Avec [detect], la zone d'illustration est lue dans le quadrilatère détecté
    plutôt que dans le rectangle centré. Rend aussi le nombre de photos où la
    détection a renoncé — celles-là retombent sur le cadrage centré, comme le
    fera l'application.
    """
    rng = random.Random(seed)
    prints = sample_prints(cards)
    distances: dict[str, list[int]] = {shot.name: [] for shot in REGIMES}
    gave_up = 0

    with httpx.Client(headers={"User-Agent": USER_AGENT}) as client:
        for index, (_, art_url, expected) in enumerate(prints, start=1):
            card = fetch_card_image(art_url, client)
            if card is None:
                continue
            for shot in REGIMES:
                photo = compose(card, shot, rng)
                quad = find_card(photo) if detect else None
                if quad is None:
                    if detect:
                        gave_up += 1
                    art = crop_art(crop_to_card_frame(photo))
                else:
                    art = sample_art(photo, quad, ART_BOX)
                distances[shot.name].append(hamming_distance(dhash(art), expected))
            print(f"  {index}/{len(prints)}", end="\r", flush=True)

    print(" " * 20, end="\r")
    return distances, gave_up


def report(distances: dict[str, list[int]], gave_up: int = 0) -> None:
    print(f"\nSeuil de confiance : {CONFIDENCE} bits")
    if gave_up:
        print(f"détections abandonnées (repli sur le cadrage centré) : {gave_up}")
    print()
    print(f"{'régime':24} {'médiane':>8} {'reconnues':>11} {'perdues':>9}")
    for shot in REGIMES:
        values = sorted(distances[shot.name])
        if not values:
            continue
        median = values[len(values) // 2]
        found = sum(1 for v in values if v <= CONFIDENCE)
        print(
            f"{shot.name:24} {median:8} "
            f"{found:>6}/{len(values):<4} {len(values) - found:>9}"
        )
    print()
    for shot in REGIMES:
        print("  " + shot.label())


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--cards", type=int, default=40)
    parser.add_argument(
        "--detect",
        action="store_true",
        help="lire l'illustration dans le quadrilatère détecté",
    )
    args = parser.parse_args()

    mode = "détection des bords" if args.detect else "cadrage centré"
    print(f"Banc de cadrage — {args.cards} cartes × {len(REGIMES)} régimes — {mode}")
    distances, gave_up = measure(args.cards, detect=args.detect)
    report(distances, gave_up)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: app/tool/framing_bench.dart
/// Banc de cadrage, côté Dart : que coûte une photo prise à main levée ?
///
/// **Jumeau de `api/app/measure/framing_bench.py`, et pourquoi il en faut un.**
/// Le banc Python mesure la détection de bords du jumeau Python. Or le code qui
/// tourne sur le téléphone est celui-ci, et c'est lui qu'il faut départager
/// quand plusieurs approches de détection sont en concurrence. Porter quatre
/// approches vers Python multiplierait les occasions de se tromper ; porter le
/// banc une fois ne les multiplie pas.
///
/// **Les photos sont synthétiques, et c'est un choix.** Une photo réelle porte
/// sa vérité terrain dans la tête de celui qui l'a prise ; une photo composée la
/// porte dans ses paramètres. On sait exactement de combien la carte a été
/// décalée, tournée, agrandie — donc à quel écart correspond quel échec. C'est
/// le seul moyen de mesurer une détection sans posséder mille cartes.
///
/// **Ce que ce banc ne prouve pas.** Une photo composée n'a ni flou de bougé, ni
/// mise au point ratée, ni reflet sur un protège-carte ; sa table est un grain
/// synthétique traversé d'un dégradé, pas un plateau à lames de bois. Un gain
/// mesuré ici est une condition nécessaire, jamais suffisante : la seule preuve
/// reste une carte de papier devant l'objectif. Le banc sert à **éliminer** ce
/// qui échoue et à comparer ce qui passe, pas à décréter qu'une approche marche.
///
/// **Le tirage vient de la base**, exporté par
/// `python -m app.measure.export_framing_set`. Même filtre et même ordre que le
/// banc Python : deux bancs qui mesureraient des cartes différentes ne seraient
/// pas comparables.
///
/// Les images sont mises en cache sur le disque. Scryfall demande un débit bas
/// et un `User-Agent` descriptif (`CLAUDE.md` §IV.4) : le cache fait qu'un banc
/// rejoué n'émet aucune requête, ce qui compte quand on le rejoue une fois par
/// approche.
///
/// Usage :
/// ```
/// dart run tool/framing_bench.dart              # détection des bords
/// dart run tool/framing_bench.dart --centered   # cadrage centré, pour comparer
/// dart run tool/framing_bench.dart --cards 12   # échantillon plus petit
/// ```
library;

import 'dart:convert';
import 'dart:io';
import 'dart:math' as math;

import 'package:deckhand/src/features/scan/domain/art_box.dart';
import 'package:deckhand/src/features/scan/domain/art_hash.dart';
import 'package:deckhand/src/features/scan/domain/card_bounds.dart';
import 'package:deckhand/src/features/scan/domain/card_framing.dart';
import 'package:deckhand/src/features/scan/domain/card_geometry.dart';
import 'package:image/image.dart' as img;

import 'synthetic_photo.dart';

/// Seuil de confiance du scan, en bits. Au-delà, la carte est « perdue » : ce
/// n'est pas une erreur silencieuse — l'application dit son doute —, mais elle
/// ne reconnaît rien.
const int confidence = 12;

Future<void> main(List<String> args) async {
  final centered = args.contains('--centered');
  final limit = int.tryParse(_option(args, '--cards') ?? '') ?? 40;

  final root = File.fromUri(Platform.script).parent;
  final setFile = File('${root.path}/${_option(args, '--set') ?? 'framing_set.json'}');
  if (!setFile.existsSync()) {
    stderr.writeln(
      'tirage absent : ${setFile.path}\n'
      'lancer d\'abord : cd api && python -m app.measure.export_framing_set',
    );
    exitCode = 66;
    return;
  }

  final entries = (jsonDecode(setFile.readAsStringSync()) as List)
      .cast<Map<String, dynamic>>()
      .take(limit)
      .toList();

  final cache = Directory('${root.path}/.framing_cache')
    ..createSync(recursive: true);

  stdout.writeln(
    'Banc de cadrage — ${entries.length} cartes × ${regimes.length} régimes — '
    '${centered ? 'cadrage centré' : 'détection des bords'}',
  );

  final distances = {for (final shot in regimes) shot.name: <int>[]};
  var gaveUp = 0;
  var skipped = 0;

  for (var i = 0; i < entries.length; i++) {
    final entry = entries[i];
    final card = await _cardImage(entry, cache);
    if (card == null) {
      skipped++;
      continue;
    }
    final expected = ArtHash.fromHex(entry['hash'] as String);
    // **Le tirage dit ce qu'il contient.** Le gabarit et l'orientation se
    // déduisent du jeu et de la disposition, au lieu d'être supposés : c'est ce
    // qui permet de mesurer les cartes couchées, que le banc ignorait.
    final frame = _frameOf(entry);
    final couchee = frame == CardFrame.riftboundWide;
    final game = entry['game'] as String? ?? 'magic';
    // La carte synthétique est découpée aux proportions du jeu tiré, et non à
    // celles de Magic : composer une carte au mauvais format mesurerait la
    // détection sur un objet qui n'existe pas.
    final aspect = cardAspectFor(game);

    for (final shot in regimes) {
      // Une graine par (carte, régime) : le grain de la table et le tirage sont
      // ainsi identiques d'une exécution à l'autre, donc d'une approche à
      // l'autre. Sans cela, deux mesures différeraient par leur bruit.
      final photo = compose(
        card,
        shot,
        math.Random(20260810 + i * 17),
        couchee: couchee,
        aspect: aspect,
      );
      final quad = centered ? null : findCard(photo, game: game);
      final img.Image art;
      if (quad == null) {
        if (!centered) gaveUp++;
        art = cropArt(cropToCardFrame(photo, game: game), frame);
      } else {
        art = sampleArt(photo, quad, frame.box);
      }
      distances[shot.name]!.add(_hamming(computeArtHash(art), expected));
    }
    stdout.write('  ${i + 1}/${entries.length}\r');
  }
  stdout.write('${' ' * 24}\r');

  _report(distances, gaveUp, skipped);
}

/// Distance de Hamming entre deux empreintes.
int _hamming(ArtHash a, ArtHash b) {
  var total = 0;
  for (var i = 0; i < a.bytes.length; i++) {
    var x = a.bytes[i] ^ b.bytes[i];
    while (x != 0) {
      total += x & 1;
      x >>= 1;
    }
  }
  return total;
}

/// Carte entière, depuis le cache disque ou depuis Scryfall.
Future<img.Image?> _cardImage(
  Map<String, dynamic> entry,
  Directory cache,
) async {
  final file = File('${cache.path}/${entry['id']}.jpg');
  if (!file.existsSync()) {
    final client = HttpClient();
    try {
      final request = await client.getUrl(Uri.parse(entry['url'] as String));
      // Descriptif, comme le garde-fou l'exige : une source doit pouvoir
      // identifier qui l'interroge et pourquoi.
      request.headers.set(
        HttpHeaders.userAgentHeader,
        'DeckHand/0.1 (https://github.com/Lelio88/DeckHand)',
      );
      final response = await request.close();
      if (response.statusCode != 200) return null;
      final bytes = await response.fold<List<int>>(
        <int>[],
        (acc, chunk) => acc..addAll(chunk),
      );
      file.writeAsBytesSync(bytes);
      // Débit bas : le banc n'a aucune raison d'être pressé, et la source en a
      // de nous limiter.
      await Future<void>.delayed(const Duration(milliseconds: 120));
    } on Object {
      return null;
    } finally {
      client.close();
    }
  }
  try {
    return img.decodeImage(file.readAsBytesSync());
  } on Object {
    return null;
  }
}

/// Fond de table texturé.
///
/// Un aplat uni rendrait la détection triviale et le banc menteur. Le grain
/// met en difficulté les approches par région ; le dégradé diagonal imite
/// l'éclairage inégal, qui est l'écueil mesuré sur une vraie photo — un coin de
/// table plus sombre passe sous le seuil de carton et fusionne avec la carte.
void _report(Map<String, List<int>> distances, int gaveUp, int skipped) {
  stdout.writeln('\nSeuil de confiance : $confidence bits');
  if (gaveUp > 0) {
    stdout.writeln(
      'détections abandonnées (repli sur le cadrage centré) : $gaveUp',
    );
  }
  if (skipped > 0) {
    stdout.writeln('cartes non téléchargées : $skipped');
  }
  stdout.writeln('');
  stdout.writeln(
    '${'régime'.padRight(24)}${'médiane'.padLeft(8)}'
    '${'reconnues'.padLeft(12)}${'perdues'.padLeft(10)}',
  );
  for (final shot in regimes) {
    final values = [...distances[shot.name]!]..sort();
    if (values.isEmpty) continue;
    final median = values[values.length ~/ 2];
    final found = values.where((v) => v <= confidence).length;
    stdout.writeln(
      '${shot.name.padRight(24)}${median.toString().padLeft(8)}'
      '${'$found/${values.length}'.padLeft(12)}'
      '${(values.length - found).toString().padLeft(10)}',
    );
  }
}

/// Gabarit à appliquer, d'après ce que le tirage déclare.
///
/// Le banc mesurait Magic seul et prenait `modern` pour acquis. Un tirage porte
/// désormais son jeu et sa disposition, ce qui est la seule façon de mesurer un
/// gabarit qui n'est pas celui par défaut — à commencer par les cartes
/// couchées, dont le cadre est mesuré depuis longtemps sans avoir jamais servi.
/// Le cadre qu'annonce une entrée du tirage.
///
/// **Se tromper ici ne se voit pas dans les nombres.** Un cadre d'un autre jeu
/// découpe de travers et rend des distances énormes, qu'on imputerait au
/// gabarit mesuré plutôt qu'au banc. C'est pourquoi le tirage porte son jeu et
/// sa disposition plutôt que de les laisser deviner.
CardFrame _frameOf(Map<String, dynamic> entry) {
  final game = entry['game'] as String? ?? 'magic';
  final layout = entry['layout'] as String? ?? 'normal';
  return switch (game) {
    'riftbound' =>
      layout == 'landscape' ? CardFrame.riftboundWide : CardFrame.riftbound,
    // `layout` porte ici le `frameType` de la source, et « pendulum » y figure
    // pour les 390 cartes dont l'illustration déborde — et pour elles seules.
    'yugioh' =>
      layout.contains('pendulum') ? CardFrame.yugiohPendulum : CardFrame.yugioh,
    _ => CardFrame.modern,
  };
}

String? _option(List<String> args, String name) {
  final i = args.indexOf(name);
  return i >= 0 && i + 1 < args.length ? args[i + 1] : null;
}
